### PR TITLE
Trainer

### PR DESCRIPTION
**Pull request by mousyball**
[feat]
* Add optimizer builder
* Add scheduler builder
* Add set_random_seed
* Add utils - create_work_dir

[refactor]
* Extract dependency in trainer
  * Move 'set_random_seed', 'get_device', 'create_work_dir' to trainer.utils

### CODE REVIEW

File: examples/train_cifar10.py [hunk 1]
```diff
@@ -85,28 +89,52 @@ def dataloader():
     config.merge_from_file('configs/pytorch_trainer/trainer.yaml')
     config.merge_from_list(['HOOK.CheckpointHook.interval', 3])
 
+    # create working directory
+    work_dir = './dev/trainer/'
+    work_dir = create_work_dir(work_dir)
```
Comment (JanLin0817) [inline examples/train_cifar10.py:94]: Why we place `create_work_dir` out of `EpochBasedTrainer`

File: examples/train_cifar10.py [hunk 2]
```diff
@@ -85,28 +89,52 @@ def dataloader():
...
+
+    # logger
+    logger = get_logger(work_dir)
```
Comment (JanLin0817) [inline examples/train_cifar10.py:97]: Why we place `get_logger` out of `EpochBasedTrainer`

File: examples/train_cifar10.py [hunk 3]
```diff
@@ -85,28 +89,52 @@ def dataloader():
...
+    set_random_seed(logger,
+                    seed=seed,
+                    deterministic=deterministic)
+
+    # get device
+    gpu_ids = config.trainer.gpu_ids
+    device = get_device(logger,
+                        gpu_ids=gpu_ids,
+                        deterministic=deterministic)
+
```
Comment (JanLin0817) [inline examples/train_cifar10.py:111]: Why we place `create_work_dir`, `get_logger`, `set_random_seed`, `get_device` out of `EpochBasedTrainer`

**Review by JanLin0817 — COMMENTED** [review]
Please review the question

File: examples/train_cifar10.py (hunk 3 above)
Comment (mousyball) [reply examples/train_cifar10.py:111]: Dependency Injection

**Pull request merged**